memparse.go

```diff
@@ -0,0 +1,181 @@
+				return err
+			}
+		}
+		return nil
+	}
+
+	for _, task := range tasks {
+		// Clear the table before processing each checkpoint task
+		table.ClearRows()
+
+		c := crit.New(nil, nil, filepath.Join(task.outputDir, "checkpoint"), false, false)
+		psTree, err := c.ExplorePs()
+		if err != nil {
+			return fmt.Errorf("failed to get process tree: %w", err)
+		}
+
+		// Populate the table rows
+		if err := traverseTree(psTree, task.outputDir); err != nil {
+			return err
+		}
+
+		fmt.Printf("\nDisplaying processes memory sizes from %s\n\n", task.checkpointFilePath)
+		table.Render()
+	}
+
+	return nil
+}
+
+func printProcessMemoryPages(task task) error {
+	memReader, err := crit.NewMemoryReader(
+		filepath.Join(task.outputDir, metadata.CheckpointDirectory),
+		pID, pageSize,
+	)
+	if err != nil {
+		return err
+	}
+
+	// Write the output to stdout by default
+	var output io.Writer = os.Stdout
+	var compact bool
+
+	if outputFilePath != "" {
+		// Write to ouput to file if --output is provided
+		output, err = os.Create(outputFilePath)
+		if err != nil {
+			return err
+		}
+	} else {
+		compact = true // Use a compact format when writing the output to stdout
+		fmt.Printf("\nDisplaying memory pages content for Process ID %d from checkpoint: %s\n\n", pID, task.checkpointFilePath)
+	}
+
+	fmt.Fprintf(output, "Address           Hexadecimal                                       ASCII            \n")
+	fmt.Fprintf(output, "-------------------------------------------------------------------------------------\n")
+
+	pagemapEntries := memReader.GetPagemapEntries()
+	for _, entry := range pagemapEntries {
+		start := entry.GetVaddr()
+		end := start + (uint64(pageSize) * uint64(entry.GetNrPages()))
+		buf, err := memReader.GetMemPages(start, end)
+		if err != nil {
+			return err
+		}
+
+		hexdump(output, buf, start, compact)
+	}
+	return nil
+}
+
+// hexdump generates a hexdump of the buffer 'buf' starting at the virtual address 'start'
+// and writes the output to 'out'. If compact is true, consecutive duplicate rows will be represented
+// with an asterisk (*).
+func hexdump(out io.Writer, buf *bytes.Buffer, vaddr uint64, compact bool) {
+	prevAscii := ""
+	duplicate := 0
+	for buf.Len() > 0 {
+		row := buf.Next(chunkSize)
+		hex, ascii := generateHexAndAscii(row)
+
+		if compact {
+			if prevAscii == ascii {
+				if duplicate == 1 {
+					fmt.Fprint(out, "*\n")
+				}
+				duplicate++
+			} else {
+				printHexdumpRow(out, vaddr, hex, ascii)
+				duplicate = 0
+			}
+		} else {
+			printHexdumpRow(out, vaddr, hex, ascii)
+		}
+
+		vaddr += chunkSize
+		prevAscii = ascii
+	}
+}
+
+func printHexdumpRow(out io.Writer, vaddr uint64, hex string, ascii string) {
+	fmt.Fprintf(out, "%016x  %s |%s|\n", vaddr, hex, ascii)
+}
+
+// generateHexAndAscii takes a byte slice and generates its hexadecimal and ASCII representations.
+func generateHexAndAscii(data []byte) (hex, ascii string) {
+	s := string(data)
```

Comment:
This cast is not required, since we directly operate on the ASCII values of the characters, which is provided by the byte slice. The function will work identically if `data` is used in place of `s`.